facet/tests/derive.rs

```diff
@@ -288,6 +288,70 @@ fn tuple_struct_field_doc_comment() {
     );
 }
 
+#[test]
+fn record_struct_generic() {
+    #[derive(Clone, Hash, PartialEq, Eq, ::facet::Facet)]
+    struct Blah<'a, T: Facet, const C: usize = 3>
+    where
+        T: core::hash::Hash,
+    {
+        field: core::marker::PhantomData<&'a T>,
+    }
+}
+
+#[test]
+fn tuple_struct_generic() {
+    #[derive(Clone, Hash, PartialEq, Eq, ::facet::Facet)]
+    #[repr(transparent)]
+    struct Blah<'a, T: Facet, const C: usize = 3>(T, core::marker::PhantomData<&'a ()>)
+    where
+        T: core::hash::Hash;
+}
+
+#[test]
+fn unit_struct_generic() {
+    #[derive(Clone, Hash, PartialEq, Eq, ::facet::Facet)]
+    struct Blah<const C: usize = 3>
+    where
+        (): core::hash::Hash;
+}
+
+#[test]
+fn enum_generic() {
+    #[allow(dead_code)]
+    #[derive(Clone, Hash, PartialEq, Eq, ::facet::Facet)]
+    #[repr(u8)]
+    enum E<'a, T: Facet, const C: usize = 3>
+    where
+        T: core::hash::Hash,
+    {
+        Unit,
+        Tuple(T, core::marker::PhantomData<&'a ()>),
+        Record {
+            field: T,
+            phantom: core::marker::PhantomData<&'a ()>,
+        },
+    }
+}
+
+// #[test]
+// fn enum_generic_partial() {
+//     #[allow(dead_code)]
+//     #[derive(Clone, Hash, PartialEq, Eq, ::facet::Facet)]
+//     #[repr(u8)]
+//     enum E<'a, T: Facet, const C: usize = 3>
+//     where
+//         T: core::hash::Hash,
+//     {
+//         Unit,
+//         Tuple(i32),
+//         Record {
+//             field: T,
+//             phantom: core::marker::PhantomData<&'a ()>,
+//         },
+//     }
+// }
```

Comment:
There is one limitation right now which is that the generated shadow structs assume all generics are used for each variant, so this currently fails to compile. We'll likely need to filter the generics and where clauses by usages of them within the variant fields. But that's kind of complicated and only syntactic which means it wont even work all the time when more macros are involved.
So I am leaving that out for now